botocore/signers.py

```diff
@@ -213,6 +220,126 @@ def generate_presigned_url(self, request_dict, expires_in=3600,
+
+        :rtype: str
+        :return: The signed URL.
+        """
+        if (date_less_than is not None and policy is not None
+                or date_less_than is None and policy is None):
+            raise ValueError('Need to provide either date_less_than or policy')
+        if policy is None and date_less_than is not None:
+            policy = self.build_policy(url, date_less_than)
+        if isinstance(policy, six.text_type):
+            policy = policy.encode('utf8')
+        if date_less_than is not None:
+            params = ['Expires=%s' % int(datetime2timestamp(date_less_than))]
+        else:
+            params = ['Policy=%s' % self._url_b64encode(policy).decode('utf8')]
+        signature = self.rsa_signer(policy)
+        params.extend([
+            'Signature=%s' % self._url_b64encode(signature).decode('utf8'),
+            'Key-Pair-Id=%s' % self.key_id,
+            ])
+        return url + ('&' if '?' in url else '?') + '&'.join(params)
+
+    def build_policy(self, resource, date_less_than,
+                     date_greater_than=None, ip_address=None):
+        """A helper to build policy.
+
+        :type resource: str
+        :param resource: The URL or the stream filename of the protected object
+
+        :type date_less_than: datetime
+        :param date_less_than: The URL will expire after the time has passed
+
+        :type date_greater_than: datetime
+        :param date_greater_than: The URL will not be valid until this time
+
+        :type ip_address: str
+        :param ip_address: Use 'x.x.x.x' for an IP, or 'x.x.x.x/x' for a subnet
+
+        :rtype: str
+        :return: The policy in a compact string.
+        """
+        if date_greater_than is None and ip_address is None:
+            policy = (
+                '{"Statement":[{"Resource":"%s",'
+                '"Condition":{"DateLessThan":{"AWS:EpochTime":%s}}}]}') % (
+                resource, int(datetime2timestamp(date_less_than)))
+        else:
+            # SEE: http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-creating-signed-url-custom-policy.html
+            moment = int(datetime2timestamp(date_less_than))
+            condition = {"DateLessThan": {"AWS:EpochTime": moment}}
+            if ip_address:
+                if '/' not in ip_address:
+                    ip_address += '/32'
+                condition["IpAddress"] = {"AWS:SourceIp": ip_address}
+            if date_greater_than:
+                moment = int(datetime2timestamp(date_greater_than))
+                condition["DateGreaterThan"] = {"AWS:EpochTime": moment}
+            custom_policy = {
+                "Statement": [{"Resource": resource, "Condition": condition}]}
+            policy = json.dumps(
+                custom_policy,
+                sort_keys=True,  # Make it stable
+                separators=(',', ':'))
+        LOG.debug("CloudFront policy: %s", policy)
```

Comment:
Do we need to log this as this is just a helper method and it is not used in flow of generating a signed url? I could see us logging it in the CLI though if we try to build on top of it.